### PR TITLE
LPS-174317 Use ClayDatePicker instead of custom DatePicker

### PR DESCRIPTION
**Pull request by Stephy6**
Motivation
=======
In the Segmentation Unification initiative there are many changes we want to make in order to improve the Segments Editor component. There are new properties in Segments Editor (derived from the unification with Analytics  Cloud) that need to use ranges of dates which we don't support. In order to do it we're migrating the date picker to the ClayDataPicker developed in clay. Besides for the properties that use data ranges we're also migrating the date picker to the ClayDatePicker for the properties that uses dates (not only date ranges).
[Link to the task](https://issues.liferay.com/browse/LPS-174317)
[Link to the story](https://issues.liferay.com/browse/LPS-173418)

Proposed Solution
========
This PR contains the changes required to use de the ClayDatePicker instead of our custom date picker. It unifies all the date properties so they use only one react component. For the ClayDatePicker we're using its properties expanded/OnExpandedChange to handle the input aside for the OnBlur property. it also checks that the input value has changed before saving the data. 

Steps to Verify:
----------------
1. Go to Segments -> People
2. Click on add button (+) to create a segment
3. Drag a property of Date type (Date Modified, Date of Birth). Check it from different groups of properties (User, Organization, Session)
4. Click on the calendar button and choose a date.
5. Save it and open the Segment again. Check that the Date is the previously saved. 
6. Put a date using the keyboard using the correct format and save it.
7. Check that the date is saved correctly. If the format is incorrect it saves the previous correct one.

*This PR doesn't contains any tests, they will be added in another one.

Screenshots:
----------------
Screenshot 2023-02-02 at 17 48 04

### CODE REVIEW

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:tango

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateInput.es.js [hunk 1]
```diff
@@ -51,20 +51,30 @@ class DateInput extends React.Component {
 		const date = jsDatetoYYYYMMDD(event.target.value);
 
 		if (date !== 'Invalid Date') {
-			this.setState({value: date}, () => {
```
Comment (claraizquierdo) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateInput.es.js:54]: Hi @Stephy6 , if we are not using this component, are these changes necessary? Is this component necessary at all?

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js [hunk 2]
```diff
@@ -43,65 +47,87 @@ class DateTimeInput extends React.Component {
...
+						`${Liferay.Language.get('august')}`,
+						`${Liferay.Language.get('september')}`,
+						`${Liferay.Language.get('october')}`,
+						`${Liferay.Language.get('november')}`,
+						`${Liferay.Language.get('december')}`,
+					]}
+					onBlur={this._saveDateTimeValue}
 					onChange={this._handleDateChange}
-					type="date"
+					onExpandedChange={this._handleExpandedChange}
```
Comment (claraizquierdo) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:130]: Do we need the `onExpendedChange` if we are using the `onBlur` ? I don't know if we can get rid of the `onExpendedChange`

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/CriteriaRow.es.js [hunk 3]
```diff
@@ -347,7 +346,7 @@ class CriteriaRow extends Component {
 		const inputComponentsMap = {
 			[PROPERTY_TYPES.BOOLEAN]: BooleanInput,
 			[PROPERTY_TYPES.COLLECTION]: CollectionInput,
-			[PROPERTY_TYPES.DATE]: DateInput,
+			[PROPERTY_TYPES.DATE]: DateTimeInput,
```
Comment (claraizquierdo) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/CriteriaRow.es.js:349]: We should not import DateTime component if we are not using it.

**Review by cgoncas — CHANGES_REQUESTED** [review]
I have found that the format of the date in the current Date Picker is not the correct one:

Current solution:

Captura de pantalla 2023-02-03 a las 10 36 56

Expected Solution:
Captura de pantalla 2023-02-03 a las 10 41 22

Also, I think this PR doesn't include tests, right?

**Review by claraizquierdo — CHANGES_REQUESTED** [review]
Hi @Stephy6 , great work so far. I've left some comments on the code. Also, I see commits with intermediate solutions that might be squashed into other commits. What do you think?

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js [hunk 4]
```diff
@@ -43,65 +47,87 @@ class DateTimeInput extends React.Component {
...
 
 		return (
 			<div className="criterion-input date-input">
-				<input
+				<ClayDatePicker
 					aria-label={`${propertyLabel}: ${Liferay.Language.get(
 						'select-date'
 					)}`}
-					className="form-control"
-					data-testid="date-input"
```
Comment (cgoncas) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:100]: If we delete the data-testid the QA tests will break, right? 

@CarlosBrichete

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js [hunk 5]
```diff
@@ -43,65 +47,87 @@ class DateTimeInput extends React.Component {
...
 
 	render() {
-		const {value} = this.state;
+		const {expanded, value} = this.state;
 		const {disabled, propertyLabel} = this.props;
 
 		return (
 			<div className="criterion-input date-input">
-				<input
+				<ClayDatePicker
```
Comment (cgoncas) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:108]: In order to set the right format I think we need to initialize the _**dateFormat**_ correctly.

**Comment by Stephy6** [comment]
> I have found that the format of the date in the current Date Picker is not the correct one:
> 
> Current solution:
> 
> Captura de pantalla 2023-02-03 a las 10 36 56
> 
> Expected Solution: Captura de pantalla 2023-02-03 a las 10 41 22
> 
> Also, I think this PR doesn't include tests, right?

Yes, I mentioned it in the description, I'm going to added now

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js (hunk 5 above)
Comment (Stephy6) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:108]: True!

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js (hunk 4 above)
Comment (Stephy6) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:100]: I'm not really sure, that property is used in the front-end tests

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/CriteriaRow.es.js (hunk 3 above)
Comment (Stephy6) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/CriteriaRow.es.js:349]: True, thanks

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js (hunk 2 above)
Comment (Stephy6) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:130]: Yes we need it because when we open the calendar the focus doesn't go to the input. The onBlur is for the cases the user sets a value using the keyboard and the OnExpendedChange is for when the user sets the value by opening the calendar and selecting

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateInput.es.js (hunk 1 above)
Comment (Stephy6) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateInput.es.js:54]: Hmm I thought I erased, thank u

**Comment by Stephy6** [comment]
ci:test:relevant

**Comment by Stephy6** [comment]
ci:test:relevant

**Comment by CarlosBrichete** [comment]
ci:stop

**Comment by CarlosBrichete** [comment]
ci:test:relevant

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js [hunk 6]
```diff
@@ -12,29 +12,33 @@
...
 
+import ClayDatePicker from '@clayui/date-picker';
 import dateFns from 'date-fns';
 import propTypes from 'prop-types';
 import React from 'react';
 
-import {PROPERTY_TYPES} from '../../utils/constants.es';
-
-const INPUT_DATE_FORMAT = 'YYYY-MM-DD';
+const INPUT_DATE_FORMAT = 'YYYY/MM/DD';
```
Comment (claraizquierdo) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:20]: Why do you use this input date format for visualizing the dates? In Figma file I see the format YYYY-MM-DD and looks like the default format for the ClayDatePicker too.

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js [hunk 7]
```diff
@@ -43,65 +47,82 @@ class DateTimeInput extends React.Component {
...
+										.toISOString()
+								: dateFns.format(
+										this.state.value,
+										OUTPUT_DATE_FORMAT
+								  ),
 					});
 				}
 			);
 		}
+		this.previousValue = dateInput;
```
Comment (claraizquierdo) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:92]: Maybe you can store this previous value in the state and update it in the same setState call. This way the code will look cleaner.

**Review by claraizquierdo — COMMENTED** [review]
Hi @Stephy6, everything seems to be working fine. I just added two comments/suggestions.

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js (hunk 7 above)
Comment (Stephy6) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:92]: Understood, thank you

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js (hunk 6 above)
Comment (Stephy6) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:20]: Ohh last time I check it was with '/'.. I'll change it

**Comment by cgoncas** [comment]
Hey @CarlosBrichete and @rafaellajordao !!

Could you take a look to the failing tests?

Thank you

**Comment by Stephy6** [comment]
ci:test:relevant

**Comment by Stephy6** [comment]
> Hey @CarlosBrichete and @rafaellajordao !!
> 
> Could you take a look to the failing tests?
> 
> Thank you

They are failing because of the years available since there are tests that are using old dates and this has to be indicated to the ClayDatePicker

**Comment by Stephy6** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
> ci:test:relevant

Hey @Stephy6 we should not run ci:test:relevant because Poshi tests are still broken, right?

**Comment by Stephy6** [comment]
> > ci:test:relevant
> 
> Hey @Stephy6 we should not run ci:test:relevant because Poshi tests are still broken, right?

I have doubts about that, because the changes that failed yesterday were the changes that Carlos made in a commit, and I removed it, so It should work

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js [hunk 8]
```diff
@@ -60,36 +61,34 @@ class DateTimeInput extends React.Component {
...
 				},
 				() => {
 					this.props.onChange({
-						type: PROPERTY_TYPES.DATE_TIME,
-						value: dateFns
-							.parse(resetDate, INPUT_DATE_FORMAT)
-							.toISOString(),
+						type: this.props.propertyType,
+						value:
+							this.props.propertyType === 'date-time'
```
Comment (claraizquierdo) [inline modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js:82]: It could be more elegant to change the string 'date-time' for PROPERTY_TYPES.DATE_TIME. This is just a suggestion :) .

**Review by claraizquierdo — CHANGES_REQUESTED** [review]
Hi again @Stephy6 , I've seen a problem with the aria-label that needs to be fixed. The other comments are just a nice to have that could be improved later.

File: modules/apps/segments/segments-web/test/js/utils.js [hunk 9]
```diff
@@ -54,11 +55,15 @@ export function testControlledDateInput({
 
 	fireEvent.blur(element);
 
-	expect(mockOnChangeFunc.mock.calls.length).toBe(1);
-
-	expect(mockOnChangeFunc.mock.calls[0][0]).toMatchObject({
-		value: newValueOnChange,
-	});
+	if (noInputChange === true) {
```
[the remark this answers is not in the document]
Comment (Stephy6) [reply modules/apps/segments/segments-web/test/js/utils.js]: Corrected, thanks

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js [hunk 10]
```diff
@@ -35,74 +36,102 @@ class DateTimeInput extends React.Component {
...
 	render() {
-		const {value} = this.state;
+		const {expanded, value} = this.state;
 		const {disabled, propertyLabel} = this.props;
 
 		return (
 			<div className="criterion-input date-input">
-				<input
+				<ClayDatePicker
 					aria-label={`${propertyLabel}: ${Liferay.Language.get(
```
[the remark this answers is not in the document]
Comment (Stephy6) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js]: Done, thank you

File: modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js (hunk 8 above)
Comment (Stephy6) [reply modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/inputs/DateTimeInput.es.js]: true, thanks

**Comment by Stephy6** [comment]
ci:test:sf

**Comment by Stephy6** [comment]
ci:test:relevant

**Comment by Stephy6** [comment]
ci:stop

**Comment by Stephy6** [comment]
ci:test:sf

**Comment by Stephy6** [comment]
ci:test:relevant

**Comment by Stephy6** [comment]
ci:test:relevant

**Comment by Stephy6** [comment]
Hi @CarlosBrichete and @rafaellajordao, could you take a look at this please?

**Comment by rafaellajordao** [comment]
[Untitled_ Feb 13 2023 8_13 PM.webm](https://user-images.githubusercontent.com/49991803/218716231-49bc9c3f-57de-48a4-a819-91c374d16ee6.webm)

Hello @Stephy6 ! while i was testing her PR I found this behavior, after adding a date it changed to an earlier day, can you check that?

**Comment by Stephy6** [comment]
> Hello @Stephy6 ! while i was testing your PR I found this behavior, after adding a date it changed to an earlier day, can you check that?

Wow, thanks a lot @rafaellajordao

**Review by claraizquierdo — CHANGES_REQUESTED** [review]
Hi @Stephy6 , for the problem that @rafaellajordao detected you can have a look at the timezone property here: https://clayui.com/docs/components/date-picker/api.html

Also, the aria labels for buttonChooseDate and buttonDot are not correct. I think switching them fix it :)

**Comment by CarlosBrichete** [comment]
ci:reevaluate:1377401_8983

**Comment by claraizquierdo** [comment]
Hi @Stephy6 , I've added a commit with a small refactor of the DateTimeInput component. Let me know your thoughts.

I've seen that the edit mode is failing when calling the `dateToInternationalHuman` method. Could you have a look at that error?

**Comment by Stephy6** [comment]
> [master] ci:test:relevant - liferay-tango > liferay-tango - PR#3212 - 2023-02-14[11:27:29]

great, thank u!

**Comment by Stephy6** [comment]
ci:test:sf

**Comment by Stephy6** [comment]
ci:test:relevant

**Comment by Stephy6** [comment]
ci:test:relevant

**Comment by Stephy6** [comment]
ci:forward
